chromium_src/components/content_settings/core/common/cookie_settings_base.cc

```diff
@@ -0,0 +1,107 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "components/content_settings/core/common/cookie_settings_base.h"
+
+#include "base/containers/flat_map.h"
+#include "base/no_destructor.h"
+#include "base/optional.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "components/content_settings/core/common/content_settings.h"
+#include "components/content_settings/core/common/content_settings_pattern.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+
+namespace content_settings {
+
+namespace {
+
+constexpr char kWp[] = "https://[*.]wp.com/*";
+constexpr char kWordpress[] = "https://[*.]wordpress.com/*";
+constexpr char kPlaystation[] = "https://[*.]playstation.com/*";
+constexpr char kSonyentertainmentnetwork[] =
+    "https://[*.]sonyentertainmentnetwork.com/*";
+
+bool BraveIsAllowedThirdParty(
+    const GURL& url,
+    const GURL& site_for_cookies,
+    const base::Optional<url::Origin>& top_frame_origin) {
+  static const base::NoDestructor<
+      // url -> first_party_url allow map
+      std::vector<std::pair<ContentSettingsPattern,
+                            ContentSettingsPattern>>> entity_list({
+          {
+            ContentSettingsPattern::FromString(kWp),
+            ContentSettingsPattern::FromString(kWordpress)
+          },
+          {
+            ContentSettingsPattern::FromString(kWordpress),
+            ContentSettingsPattern::FromString(kWp)
+          },
+          {
+            ContentSettingsPattern::FromString(kPlaystation),
+            ContentSettingsPattern::FromString(kSonyentertainmentnetwork)},
+          {
+            ContentSettingsPattern::FromString(kSonyentertainmentnetwork),
+            ContentSettingsPattern::FromString(kPlaystation)
+          },
+      });
+
+  GURL first_party_url = site_for_cookies;
+
+  if (!first_party_url.is_valid() && top_frame_origin)
```

Comment:
hmm, shouldn't we use top_frame_origin by default and fallback to site_for_cookies?

Reply:
It doesn't really matter which way you do it because if the parent frame is 3rd-party then the site_for_cookies is invalid
```// Make sure every ancestors are same-domain with the main document. Otherwise
  // this will be a 3rd party cookie.
  for (const RenderFrameHostImpl* rfh = render_frame_host; rfh;
       rfh = rfh->parent_) {
    if (!candidate.IsEquivalent(
            net::SiteForCookies::FromOrigin(rfh->last_committed_origin_))) {
      return net::SiteForCookies();
    }
  }```

Reply:
and it seems safer to me to use site_for_cookies when it is available because that is its purpose

Reply:
I just want to be consistent with how the original class thinks about first party url
```
  GetCookieSettingInternal(
      url, top_frame_origin ? top_frame_origin->GetURL() : site_for_cookies,
      false, nullptr, &setting);
```

Reply:
@iefremov so the problem here is that site_for_cookies is treated differently in default 3rd-party cookie blocking and in content settings. The content settings default to the top_frame_origin, but the 3rd-party cookie settings default to the site_for_cookies. The reason we don't want to default to the top_frame_origin here is that there are times when it's not what we want (like redirects) and that is why site_for_cookies exists in the first place. We used only top frame origin in browser-laptop and there were times when it was wrong because of that.

Reply:
so for the purposes of 3rd-party cookie blocking we want to use site_for_cookies whenever it's available and only use top_frame_origin when it's empty (3rd-party iframe)

Reply:
in the folow-up I'll add a comment here explaining that and see if I can come up with a test for it